homeassistant/components/sensor/forecast.py

```diff
@@ -129,75 +122,72 @@ def unit_of_measurement(self):
     @property
     def unit_system(self):
         """Return the unit system of this entity."""
-        return self._unit_system
+        return self.forecast_data.unit_system
+
+    def update_unit_of_measurement(self):
+        """Update units based on unit system."""
+        unit_index = {
+            'si': 1,
+            'us': 2,
+            'ca': 3,
+            'uk': 4,
+            'uk2': 5
+        }.get(self.unit_system, None)
+        self._unit_of_measurement = SENSOR_TYPES[self.type][unit_index]
 
-    # pylint: disable=too-many-branches,too-many-statements
     def update(self):
         """Get the latest data from Forecast.io and updates the states."""
-        import forecastio
+        # Call the API for new forecast data. Each sensor will re-trigger this
+        # same exact call, but thats fine. We cache results for a short period
+        # of time to prevent hitting API limits. Note that forecast.io will
+        # charge users for too many calls in 1 day, so take care when updating.
+        self.forecast_data.update()
+        self.update_unit_of_measurement()
+
+        if self.type == 'minutely_summary':
+            self.forecast_data.update_minutely()
+            minutely = self.forecast_data.data_minutely
+            self._state = getattr(minutely, 'summary', '')
+        elif self.type == 'hourly_summary':
+            self.forecast_data.update_hourly()
+            hourly = self.forecast_data.data_hourly
+            self._state = getattr(hourly, 'summary', '')
+        elif self.type == 'daily_summary':
+            self.forecast_data.update_daily()
+            daily = self.forecast_data.data_daily
+            self._state = getattr(daily, 'summary', '')
+        else:
+            self.forecast_data.update_currently()
+            currently = self.forecast_data.data_currently
+            self._state = self.get_currently_state(currently)
 
-        self.forecast_client.update()
+    def get_currently_state(self, data):
+        """
+        Helper function that returns a new state based on the type.
 
-        try:
-            if self.type == 'minutely_summary':
-                self.forecast_client.update_minutely()
-                self._state = self.forecast_client.data_minutely.summary
-                return
+        If the sensor type is unknown, the current state is returned.
+        """
+        lookup_type = convert_to_camel(self.type)
+        state = getattr(data, lookup_type, 0)
 
-            elif self.type == 'hourly_summary':
-                self.forecast_client.update_hourly()
-                self._state = self.forecast_client.data_hourly.summary
-                return
+        # Some state data needs to be rounded to whole values or converted to
+        # percentages
+        if self.type in ['precip_probability', 'cloud_cover', 'humidity']:
+            return round(state * 100, 1)
+        elif (self.type in ['dew_point', 'temperature', 'apparent_temperature',
```

Comment:
If you don't need to change the lists, consider using tuples instead. Nice solution overall though. :+1: